jax/tree_util.py

```diff
@@ -172,28 +171,66 @@ def __ne__(self, other):
 
 
 class PyLeaf(object):
+  include_subclasses = True
   def __repr__(self):
     return '*'
 
 leaf = PyLeaf()
 
 def dict_to_iterable(xs):
-  keys = tuple(sorted(xs.keys()))
-  return tuple(map(xs.get, keys)), keys
+  keys, values = zip(*sorted(xs.items()))
+  return values, keys
 
 class NodeType(object):
-  def __init__(self, name, to_iterable, from_iterable):
+  def __init__(self, name, to_iterable, from_iterable, include_subclasses):
     self.name = name
     self.to_iterable = to_iterable
     self.from_iterable = from_iterable
+    self.include_subclasses = include_subclasses
 
 node_types = {}
 
-def register_pytree_node(py_type, to_iterable, from_iterable):
+def get_node_type(py_type):
+  for cls in py_type.__mro__:
```

Comment:
Accessing `pytype.__mro__` feels kind of dangerous, like we're going from something really simple and minimal to something that knows a lot about Python inheritance. Could this be fragile? Are there alternatives? For example, it seems like using `isinstance` on all the registered node types might be an alternative; I'm not saying that's better, but I'd like to think through the solution space.

I'm also interested if this has a big impact on performance. An advantage of the `type`-keyed dict mechanism is that it should be quite fast. I'm not saying the performance _is_ much different here, or that it's actually an important difference, but we should put some thought into that. What do you think?